### PR TITLE
Web Inspector doesn't show nested workers

### PR DESCRIPTION
**Pull request by dcrousso**
#### b5d3f9b6724bf506145bf2233e392479be0a95fa
<pre>
Web Inspector doesn&apos;t show nested workers
<a href="https://bugs.webkit.org/show_bug.cgi?id=255402">https://bugs.webkit.org/show_bug.cgi?id=255402</a>
&lt;<a href="https://rdar.apple.com/problem/108322385">rdar://problem/108322385</a>&gt;

Reviewed by Alexey Proskuryakov.

This allows for sub-`Worker` to appear in Web Inspector in the same way as the parent `Worker`.

* Source/JavaScriptCore/inspector/protocol/Worker.json:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::createLazyAgents):
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::createLazyAgents):
* Source/WebCore/inspector/agents/InspectorWorkerAgent.h:
* Source/WebCore/inspector/agents/InspectorWorkerAgent.cpp:
(WebCore::InspectorWorkerAgent::InspectorWorkerAgent):
(WebCore::InspectorWorkerAgent::enable):
(WebCore::InspectorWorkerAgent::connectToAllWorkerInspectorProxiesForPage): Deleted.
* Source/WebCore/inspector/agents/page/PageWorkerAgent.h: Added.
* Source/WebCore/inspector/agents/page/PageWorkerAgent.cpp: Added.
(WebCore::PageWorkerAgent::PageWorkerAgent):
(WebCore::PageWorkerAgent::connectToAllWorkerInspectorProxies):
* Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h: Added.
* Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.cpp: Added.
(WebCore::WorkerWorkerAgent::WorkerWorkerAgent):
(WebCore::WorkerWorkerAgent::connectToAllWorkerInspectorProxies):
Add support for the `Worker` inspector protocol domain inside `&quot;worker&quot;` inspector targets.

* Source/WebCore/bindings/js/WorkerModuleScriptLoader.h:
* Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp:
(WebCore::WorkerModuleScriptLoader::load):
(WebCore::WorkerModuleScriptLoader::notifyFinished):
* Source/WebCore/fileapi/FileReaderLoader.h:
* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::didReceiveResponse):
(WebCore::FileReaderLoader::didFinishLoading):
(WebCore::FileReaderLoader::didFail):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorThreadableLoaderClient::didReceiveResponse):
(WebCore::InspectorThreadableLoaderClient::didFinishLoading):
(WebCore::InspectorThreadableLoaderClient::didFail):
* Source/WebCore/loader/DocumentThreadableLoader.cpp:
(WebCore::DocumentThreadableLoader::cancel):
(WebCore::DocumentThreadableLoader::didReceiveResponse):
(WebCore::DocumentThreadableLoader::didFinishLoading):
(WebCore::DocumentThreadableLoader::didFail):
(WebCore::DocumentThreadableLoader::preflightFailure):
(WebCore::DocumentThreadableLoader::logErrorAndFail):
* Source/WebCore/loader/ThreadableLoaderClient.h:
(WebCore::ThreadableLoaderClient::didReceiveResponse):
(WebCore::ThreadableLoaderClient::didFinishLoading):
(WebCore::ThreadableLoaderClient::didFail):
* Source/WebCore/loader/ThreadableLoaderClientWrapper.h:
(WebCore::ThreadableLoaderClientWrapper::didReceiveResponse):
(WebCore::ThreadableLoaderClientWrapper::didFinishLoading):
(WebCore::ThreadableLoaderClientWrapper::didFail):
(WebCore::ThreadableLoaderClientWrapper::didReceiveAuthenticationCancellation): Deleted.
* Source/WebCore/loader/WorkerThreadableLoader.h:
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::WorkerThreadableLoader::MainThreadBridge::cancel):
(WebCore::WorkerThreadableLoader::MainThreadBridge::didReceiveResponse):
(WebCore::WorkerThreadableLoader::MainThreadBridge::didReceiveData):
(WebCore::WorkerThreadableLoader::MainThreadBridge::didFinishLoading):
(WebCore::WorkerThreadableLoader::MainThreadBridge::didFail):
* Source/WebCore/Modules/fetch/FetchLoader.h:
* Source/WebCore/Modules/fetch/FetchLoader.cpp:
(WebCore::FetchLoader::didReceiveResponse):
(WebCore::FetchLoader::didFinishLoading):
(WebCore::FetchLoader::didFail):
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.h:
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
(WebCore::NotificationResourcesLoader::ResourceLoader::didReceiveResponse):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFinishLoading):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFail):
* Source/WebCore/page/EventSource.h:
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::didReceiveResponse):
(WebCore::EventSource::didFinishLoading):
(WebCore::EventSource::didFail):
* Source/WebCore/workers/Worker.h:
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::didReceiveResponse):
(WebCore::Worker::notifyFinished):
* Source/WebCore/workers/WorkerFontLoadRequest.h:
* Source/WebCore/workers/WorkerFontLoadRequest.cpp:
(WebCore::WorkerFontLoadRequest::didReceiveResponse):
(WebCore::WorkerFontLoadRequest::didFinishLoading):
(WebCore::WorkerFontLoadRequest::didFail):
* Source/WebCore/workers/WorkerScriptLoader.h:
* Source/WebCore/workers/WorkerScriptLoader.cpp:
(WebCore::WorkerScriptLoader::didReceiveResponse):
(WebCore::WorkerScriptLoader::didFinishLoading):
(WebCore::WorkerScriptLoader::didFail):
(WebCore::WorkerScriptLoader::notifyError):
(WebCore::WorkerScriptLoader::notifyFinished):
* Source/WebCore/workers/WorkerScriptLoaderClient.h:
* Source/WebCore/workers/service/ServiceWorkerJob.h:
* Source/WebCore/workers/service/ServiceWorkerJob.cpp:
(WebCore::ServiceWorkerJob::didReceiveResponse):
(WebCore::ServiceWorkerJob::notifyFinished):
* Source/WebCore/workers/shared/SharedWorkerScriptLoader.h:
* Source/WebCore/workers/shared/SharedWorkerScriptLoader.cpp:
(WebCore::SharedWorkerScriptLoader::didReceiveResponse):
(WebCore::SharedWorkerScriptLoader::notifyFinished):
* Source/WebCore/xml/XMLHttpRequest.h:
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::didFail):
(WebCore::XMLHttpRequest::didFinishLoading):
(WebCore::XMLHttpRequest::didReceiveResponse):
Pass along the toplevel `ScriptExecutionContextIdentifier` so that clients can use that to get the toplevel context and do things with it (e.g. &quot;notify the `Document` that this load has finished&quot;).
This is used by Web Inspector because `Worker` do not have an `InspectorNetworkAgent`, meaning that any `InspectorInstrumentation` that pipe through to it need to be done on the same thread as the `Document`.

* Source/WebCore/workers/WorkerGlobalScope.cpp:
(WebCore::WorkerGlobalScope::importScripts):
Remove redundant `InspectorInstrumentation` call.

* Source/WebInspectorUI/UserInterface/Controllers/WorkerManager.js:
(WI.WorkerManager):
(WI.WorkerManager.prototype.workerCreated):
(WI.WorkerManager.prototype.workerTerminated):
(WI.WorkerManager.prototype.dispatchMessageFromWorker):
Keep a multimap of sub-`Worker` for each `Worker`.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype.scriptDidParse):
* Source/WebInspectorUI/UserInterface/Models/Script.js:
(WI.Script.prototype.couldBeMainResource): Added.
(WI.Script.prototype._resolveResource):
`InspectorInstrumentation::didReceiveScriptResponse` is not always called in time before `Network.responseReceived` due to the fact that the `Worker` was created on another thread (i.e. sub-`Worker`), the `WI.Script` was loaded from memory cache, or some other reason. As such, we cannot rely on `WI.Resource.Type.Script` being the only way to associate a `WI.Resource` with a `WI.Script`.  In the sub-`Worker` scenario, do a second pass to see if any `WI.Resource` with the same URL happen to have a script MIME type (even though they&apos;re marked as some other type) and associate with that.

* Source/WebInspectorUI/UserInterface/Views/QuickConsole.js:
(WI.QuickConsole):
(WI.QuickConsole.prototype.closed):
(WI.QuickConsole.prototype._populateActiveExecutionContextNavigationItemContextMenu):
(WI.QuickConsole.prototype._handleTargetAdded): Added.
Show sub-`Worker` underneath and indented from the parent `Worker`.
Also refresh the execution context picker when new targets (e.g. `Worker`) are added.

* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._handleTreeSelectionDidChange):
Only use the associated `WI.Resource` when showing a `WI.Script` if it&apos;s not the main resource. Otherwise, things like `WI.WorkerTreeElement` that use the `WI.Script` as the `representedObject` won&apos;t be found when searching for a `WI.TreeElement` matching the `representedObject`.

* Source/WebCore/inspector/InspectorInstrumentation.h:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::instrumentingAgents):
(WebCore::InspectorInstrumentation::willSendRequest):
(WebCore::InspectorInstrumentation::didReceiveResourceResponse):
(WebCore::InspectorInstrumentation::didReceiveData):
(WebCore::InspectorInstrumentation::didFinishLoading):
(WebCore::InspectorInstrumentation::didFailLoading):
(WebCore::InspectorInstrumentation::didReceiveScriptResponse):
Drive-by: Mark instrumentation only used by `ServiceWorker` as such.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* LayoutTests/inspector/worker/resources/subworker-manager.js: Added.
* LayoutTests/inspector/worker/resources/worker-utilities.js:
(TestPage.registerInitializer.awaitTarget): Added.
* LayoutTests/inspector/worker/console-basic-subworker.html: Added.
* LayoutTests/inspector/worker/console-basic-subworker-expected.txt: Added.
* LayoutTests/inspector/worker/debugger-pause-subworker.html: Added.
* LayoutTests/inspector/worker/debugger-pause-subworker-expected.txt: Added.
* LayoutTests/inspector/worker/resources-in-subworker.html: Added.
* LayoutTests/inspector/worker/resources-in-subworker-expected.txt: Added.
* LayoutTests/inspector/worker/runtime-basic-subworker.html: Added.
* LayoutTests/inspector/worker/runtime-basic-subworker-expected.txt: Added.
* LayoutTests/platform/mac-site-isolation/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/279793@main">https://commits.webkit.org/279793@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/39b56e9c7d03575ac136eec7482520b1e4bef1b6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7153 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57862 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5315 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44203 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3572 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56676 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47273 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25335 "Found 1 new API test failure: /WPE/TestMultiprocess:/webkit/WebKitWebView/multiprocess-create-ready-close (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4603 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3455 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47946 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4819 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59452 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54087 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4959 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51628 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30964 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47359 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51002 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11928 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31942 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66385 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30752 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12655 "Passed tests") |

### CODE REVIEW

**Review by patrickangle — COMMENTED** [review]
Flakey test failures look real.

File: Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp [hunk 1]
```diff
@@ -100,9 +100,13 @@ void WorkerModuleScriptLoader::load(ScriptExecutionContext& context, URL&& sourc
     }
 
     if (cspCheckFailed) {
-        m_scriptLoader->notifyError();
+        // FIXME: Always get the `ScriptExecutionContextIdentifier` of the `Document`.
+        ScriptExecutionContextIdentifier mainContext;
+        if (auto* document = dynamicDowncast<Document>(context))
+            mainContext = document->identifier();
+        m_scriptLoader->notifyError(mainContext);
```
Comment (achristensen07) [inline Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp]: Why don't we introduce ScriptExecutionContext::mainContextIdentifier()?  It seems like that would fix many of the FIXMEs in this PR.

File: Source/WebCore/inspector/agents/page/PageWorkerAgent.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+class PageWorkerAgent final : public InspectorWorkerAgent {
+public:
+    PageWorkerAgent(PageAgentContext&);
+    ~PageWorkerAgent();
+
+private:
+    void connectToAllWorkerInspectorProxies() override;
+
+    Page& m_page;
```
Comment (achristensen07) [inline Source/WebCore/inspector/agents/page/PageWorkerAgent.h:42]: Can this be a CheckedRef?

File: Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2016 Apple Inc. All rights reserved.
```
Comment (achristensen07) [inline Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h]: This doesn't look correct.

File: Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class WorkerWorkerAgent final : public InspectorWorkerAgent {
+public:
+    WorkerWorkerAgent(WorkerAgentContext&);
+    ~WorkerWorkerAgent();
+
+private:
+    void connectToAllWorkerInspectorProxies() override;
+
+    WorkerOrWorkletGlobalScope& m_globalScope;
```
Comment (achristensen07) [inline Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h:42]: Can this be a WeakPtr<WorkerOrWorkletGlobalScope> ?

File: Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp (hunk 1 above)
Comment (dcrousso) [reply Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp]: are you suggesting that i just move this logic into a `ScriptExecutionContext::mainContextIdentifier()` or do something else to actually always get the `Document::identifier()`?

there are only two spots with FIXME of this nature so im not entirely sure how much benefit we'd get from that as opposed to leaving this as is here

File: Source/WebCore/inspector/agents/page/PageWorkerAgent.h (hunk 2 above)
Comment (dcrousso) [reply Source/WebCore/inspector/agents/page/PageWorkerAgent.h:42]: Web Inspector objects normally dont do any sort of reference counting as they are often destroyed inside the destructor of the relevant object (or at the very least should never outlive it).  im hesitant to break from what's been done everywhere else in this one situation, so id rather keep it as is and change all of them in the future if that's ok

File: Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h (hunk 4 above)
Comment (dcrousso) [reply Source/WebCore/inspector/agents/worker/WorkerWorkerAgent.h:42]: ditto my comment above about this being the pattern used by Web Inspector elsewhere and also that this should not ever outlive the `WorkerOrWorkletGlobalScope`

**Comment by kkinnunen-apple** [comment]
I need this often, can we get this or something simpler landed?

**Review by rcaliman-apple — COMMENTED** [review]
(Copying over from discussion on Slack)

I've been trying to understand how the flow of connecting Web Inspector to a worker works. I _think_ I understand it.

I’ll write it out here so others can verify if I missed anything important:

---

## Act 1

Whenever a worker is spawned in WebCore, corresponding `WorkerMessagingProxy` and :star:️`WorkerInspectorProxy` objects are automatically created:
https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/workers/Worker.cpp#L84

(the fact that the implementation of `WorkerGlobalScopeProxy::create()` is in the file `WorkerMessagingProxy.cpp` really tripped me up)

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/workers/WorkerMessagingProxy.cpp#L94-L102 

Once the source file for the worker is loaded (`Worker::notifyFinished()`) it announces `WorkerMessagingProxy::startWorkerGlobalScope()` -> `WorkerInspectorProxy::workerStarted()`

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/workers/Worker.cpp#L227
https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/workers/WorkerMessagingProxy.cpp#L170

A :star:️`WorkerInspectorController` is also automatically instantiated when a `WorkerGlobalScope` is created (subclassed from `WorkerOrWorkletGlobalScope`).

## Act 2

`WorkerInspectorProxy::workerStarted()` goes to `InspectorWorkerAgent::workerStarted()` via `InspectorInstrumentation::workerStartedImpl()` and `persistentWorkerAgent` getter generated in `DerivedSources/WebCore/WebCore/inspector/InstrumentingAgents.h`

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/inspector/InspectorInstrumentation.cpp#L1072-L1076

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/inspector/agents/InspectorWorkerAgent.cpp#L120

## Act 3
The magic happens in `InspectorWorkerAgent::connectToWorkerInspectorProxy()`. 
It triggers two important flows:

1) it asks the `WorkerInspectorProxy` to connect itself to `InspectorWorkerAgent` and asks :star: `WorkerInspectorController` (already created as part of `WorkerGlobalScope`) to connect the frontend. This is when relevant agents for the worker get instantiated in `WorkerInspectorController.createLazyAgents()`. Enablement comes later.

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/inspector/agents/InspectorWorkerAgent.cpp#L165-L172

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/workers/WorkerInspectorProxy.cpp#L110-L121

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebCore/inspector/WorkerInspectorController.cpp#L98-L117

2) it dispatches a notification to the frontend that a worker was created: `m_frontendDispatcher->workerCreated()`.

The frontend handles this notification in `WI.WorkerManager.workerCreated()` and creates a corresponding `WI.WorkerTarget` representation. It then initializes it.

This initialization on the frontend calls `enable()` on the agents from the backend created in the previous step.

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebInspectorUI/UserInterface/Controllers/WorkerManager.js#L45-L59

https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebInspectorUI/UserInterface/Protocol/Target.js#L74-L77


https://github.com/WebKit/WebKit/blob/f0eff1790b685946e909097fcc0109afbaa414be/Source/WebInspectorUI/UserInterface/Controllers/WorkerManager.js#L40

This completes the cycle of connecting a newly spawned worker to the Web Inspector.

File: LayoutTests/inspector/worker/resources/subworker.js
Comment (rcaliman-apple) [inline LayoutTests/inspector/worker/resources/subworker.js:1]: The name of this file is a bit misleading. It's not used as a subworker itself, but it spawns and manages subworkers.

How about renaming to `subworker-manager.js` instead?

File: LayoutTests/inspector/worker/runtime-basic-subworker.html [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+let worker = new Worker("resources/subworker.js");
+worker.postMessage({url: "worker-1.js", data: null});
+
+async function test()
+{
+    let mainTarget = WI.mainTarget;
+
+    let subworkerTarget = await window.awaitTarget((target) => target instanceof WI.WorkerTarget && target.displayName.endsWith("worker-1.js"));
+    if (!subworkerTarget) {
```
Comment (rcaliman-apple) [inline LayoutTests/inspector/worker/runtime-basic-subworker.html]: Is there ever a condition that satisfies `if(!subworkerTarget)`?

`window.awaitTarget` runs an infinite loop that only returns when the predicate returns `true`, else it keeps `await`-ing the next `WI.TargetManager.Event.TargetAdded`.

File: LayoutTests/inspector/worker/resources-in-subworker.html [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    suite.addTestCase({
+        name: "Subworker.Resource.Start",
+        description: "Start the subworker.",
+        async test() {
+            let [targetAddedEvent, subworkerIsReadyEvent] = await Promise.all([
+                WI.targetManager.awaitEvent(WI.TargetManager.Event.TargetAdded),
+                InspectorTest.awaitEvent("TestPage-SubworkerIsReady"),
+                InspectorTest.evaluateInPage("triggerSubworkerStart()"),
+            ]);
```
Comment (rcaliman-apple) [inline LayoutTests/inspector/worker/resources-in-subworker.html:66]: Neat! 👍

File: LayoutTests/inspector/worker/resources-in-subworker.html [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            InspectorTest.expectEqual(resource.type, WI.Resource.Type.XHR, "Resource should be XHR.");
+            InspectorTest.expectEqual(resource.target, subworkerTarget, "Resource should be tied to the Subworker Target.");
+            InspectorTest.expectNull(resource.parentFrame, "Resource has no parent frame.");
+            dumpWorkerResources();
+        },
+    });
+
+    suite.addTestCase({
+        name: "Main.Resource.XHR",
+        description: "Main XHRs are tied to the Main Target and not affect the Subworker Target",
```
Comment (rcaliman-apple) [inline LayoutTests/inspector/worker/resources-in-subworker.html]: NIT: 

```diff
- Main XHRs are tied to the Main Target and not affect the Subworker Target
+ Main XHRs are tied to the Main Target and do not affect the Subworker Target
```

ditto below for `Main.Resource.Fetch` test case description

File: LayoutTests/platform/mac/TestExpectations [hunk 8]
```diff
@@ -979,6 +979,7 @@ webkit.org/b/181832 inspector/unit-tests/throttler.html [ Pass Failure ]
 webkit.org/b/161361 inspector/unit-tests/yieldable-task.html [ Pass Failure ]
 webkit.org/b/169228 inspector/worker/console-basic.html [ Pass Timeout ]
 webkit.org/b/164872 inspector/worker/debugger-multiple-targets-pause.html [ Pass Failure Timeout ]
+webkit.org/b/164872 inspector/worker/debugger-pause-subworker.html [ Pass Failure Timeout ]
```
Comment (rcaliman-apple) [inline LayoutTests/platform/mac/TestExpectations]: Why do we expect the test to Fail?

**Review by rcaliman-apple — COMMENTED** [review]
Reviewed the parts related to Web Inspector. Looks good! Thank you for building this!

I have a few notes/questions about the tests, but otherwise I recommend approval by a reviewer 👍. CC @patrickangle

---

I think I understand the flow as detailed in https://github.com/WebKit/WebKit/pull/13614#pullrequestreview-1994386213

I understand the key enabler here is having the `Worker` target own other `Worker` targets with the change in `Worker.json` and the newly introduced `WorkerWorkerAgent` kicking off the instantiation of agents in `WorkerInspectorController`.

I checked-out this locally, rebased and built to test it. It works well with the test case attached to https://bugs.webkit.org/show_bug.cgi?id=272382

Nested workers show up in the Sources tab and console statements are correctly assigned to the appropriate scripts.

nested-workers-web-inspector

File: LayoutTests/inspector/worker/runtime-basic-subworker.html (hunk 5 above)
Comment (dcrousso) [reply LayoutTests/inspector/worker/runtime-basic-subworker.html]: i wanna say there was a reason for this when i first wrote this PR, but it's been so long that i cant remember, and looking at it again now it definitely seems impossible for `subworkerTarget` to be falsy, so yeah ill remove it.  good catch!

File: LayoutTests/platform/mac/TestExpectations (hunk 8 above)
Comment (dcrousso) [reply LayoutTests/platform/mac/TestExpectations]: i think it's for a similar reason as the linked bug, but i cant really remember for sure.  i can try testing it again locally

**Comment by kkinnunen-apple** [comment]
Just a friendly ping: is this pending on implementation or on review?

**Review by aproskuryakov — APPROVED** [review]
rs=me based on Razvan's comments. Please consider his renaming suggestion, and follow up on the pass/faiure/timeout test if you can.

**Pull request merged**